doc/finch.1.in
.\" Copyright (c) 2006, Sadrul Habib Chowdhury <sadrul@users.sf.net>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.TH finch 1
.SH NAME
Finch \- A Pimpin' Penguin console frontend to libpurple. Instant
Messaging client.
.SH SYNOPSIS
.TP 5
\fBfinch \fI[options]\fR

.SH DESCRIPTION
.PP
\fBfinch\fR is a console-based modular messaging client based on libpurple
which is capable of connecting to AIM, XMPP, ICQ, IRC, SILC,
Novell GroupWise, Lotus Sametime, Zephyr, Gadu-Gadu, and QQ all at once. It has
many common features found in other clients, as well as many unique features.
Finch is not endorsed by or affiliated with America Online, ICQ, or Microsoft.

.SH OPTIONS
The following options are provided by \fBfinch\fR using the standard GNU
command line syntax:
.TP
.B \-c, \-\-config=\fIDIR\fB
Use \fIDIR\fR as the directory for config files instead of \fI~/.purple\fR.
.TP
.B \-d, \-\-debug
Print debugging messages to stderr and start with the \fBDebug\fR window. The
messages shown in the \fBDebug\fR window are the same as the ones printed in
stderr.
.TP
.B \-h, \-\-help
Print this help and exit.
.TP
.B \-n, \-\-nologin
Don't automatically login when \fBfinch\fR starts.  Sets all accounts to
Offline.
.TP
.B \-v, \-\-version
Display the version information window.

.SH GNT Shortcuts
You can use the following shortcuts (see the "Widget Actions" section for a more complete list):
.TP
.B Alt \+ a
Bring up a list of available actions. You can use this list to access the
accounts window, plugins window, preference window etc.
.TP
.B Alt \+ n
Go to the next window.
.TP
.B Alt \+ p
Go to the previous window.
.TP
.B Alt \+ w
Show the list of windows. You can select and jump to any window from the list.
.TP
.B Alt \+ c
Close the current window.
.TP
.B Alt \+ q
Quit.
.TP
.B Alt \+ m
Start moving a window. Press the cursor keys to move the window. When you are
done, press \fBEnter\fR or \fBEscape\fR.
.TP
.B Alt \+ r
Start resizing a window. Press the cursor keys to resize the window. When you
are done, press \fBEnter\fR or \fBEscape\fR.
.TP
.B Alt \+ D
Dump the contents of the screen in HTML format.
.TP
.B Alt \+ .
Move the position of the current window in the window list one place to the
right.
.TP
.B Alt \+ ,
Move the position of the current window in the window list one place to the
left.
.TP
.B Alt \+ l
Refresh the windows. This is useful after resizing the terminal window.
.TP
.B Alt \+ 1 2 ... 0
Jump to the 1st, 2nd ... 10th window.
.TP
.B Alt \+ Tab
Jump to the next URGENT (highlighted) window.
.TP
.B Alt \+ Shift \+ Tab
Jump to the previous URGENT (highlighted) window.
.TP
.B Ctrl \+ o \fR or \fB F10
Bring up the menu (if there is one) for a window.
.TP
.B F11 \fR or \fB Ctrl \+ x
Popup the context menu (if there is one) for the selected widget.
.TP
.B Alt \+ /
Show a list of available key-bindings for the current widget in focus.
.TP
.B Alt \+ \>
Switch to the next workspace
.TP
.B Alt \+ \<
Switch to the previous workspace
.TP
.B Alt \+ t
Tag (or untag) the current window
.TP
.B Alt \+ T
Attached all the tagged windows to the current workspace
.TP
.B Alt \+ s
Show the workspace list
.TP
.B F9
Create a new workspace and switch to it

.SH FILES
\fI~/.gntrc\fR: configuration file for gnt applications.
.br
.TP
A sample file looks like:
.br
[Finch]
.br
color-available = green; black
.br
color-away = blue; black
.br
color-idle = gray; black
.br
color-offline = red; black
.br
color-message-sent = cyan; default
.br
color-message-received = red; default
.br
color-message-highlight = black; green
.br
color-message-action = yellow; default
.br
color-timestamp = blue; default
.br
#See below for details on color
.br

[general]
.br
shadow = 0
.br
# There is experimental mouse support
.br
mouse = 1
.br
# To use some custom window-manager
.br
wm = /usr/local/lib/gnt/s.so
.br
# There's also a custom window manager called irssi.so
.br
# Remember window-positions based on the titles (on by default)
.br
remember_position = 1
.br
# Use borderless one-line high buttons
.br
small-button = true

.br
# Workspaces are created simply by adding Workspace-X groups as follows:
.br
[Workspace-1]
.br
name = blist
.br
# window-names specifies that windows with these semi-colon separated names are placed
into this workspace
.br
window-names = buddylist;debug-window
.br

.br
[Workspace-2]
.br
name = IM
.br
window-names = conversation-window
.br
# window-titles specifies that windows with these semi-colon separated titles are placed
into this workspace. These are matched as substrings. Window titles take precedence over
names.
.br
window-titles = Preferences;Pounce
.br

.br
[colors]
.br
# The RGB values range in [0, 1000]
.br
black = 0; 0; 0
.br
red = 1000; 0; 0
.br
green = 0; 1000; 0
.br
blue = 250; 250; 700
.br
white = 1000; 1000; 1000
.br
gray = 700; 700; 700
.br
darkgray = 256; 256; 256
.br

.br
[colorpairs]
.br
normal = white; black
.br
highlight = white; blue
.br
highlightd = black; gray
.br
shadow = black; darkgray
.br
title = white; blue
.br
titled = white; gray
.br
text = white; blue
.br
disabled = gray; black
.br
urgent = green; black
.br

.br
# Remap some keys for GntEntry
.br
[GntEntry::remap]
.br
# Remap the up-arrow to the left-arrow
.br
^[[A = ^[[D
.br
# Remap the down-arrow to the right-arrow
.br
^[[B = ^[[C
.br
# Remap 'a' to 'bcd'
.br
a = bcd
.br
# Completely ignore the key 'q'
.br
q = 
.br
# But the following will NOT work
.br
#abc = bcd
.br

# Hitting 'space' will activate a button
.br
[GntButton::remap]
.br
\\  = \\r
.br

.SH Widget Actions
You can specifiy key-bindings for specific widgets. The following entries in
\fI~/.gntrc\fR correspond to the default keybindings for the actions:

.br
[GntBox::binding]
.br
tab = focus-next
.br
right = focus-next
.br
left = focus-prev

.br
[GntComboBox::binding]
.br
down = dropdown
.br
up = dropdown

.br
[GntEntry::binding]
.br
c-a = cursor-home
.br
home = cursor-home
.br
c-e = cursor-end
.br
end = cursor-end
.br
backspace = delete-prev
.br
del = delete-next
.br
c-d = delete-next
.br
c-u = delete-start
.br
c-k = delete-end
.br
c-b = cursor-prev
.br
left = cursor-prev
.br
c-f = cursor-next
.br
right = cursor-next
.br
tab = suggest-show
.br
down = suggest-next
.br
up = suggest-prev
.br
page-down = suggest-next-page
.br
page-up = suggest-prev-page
.br
c-w = delete-prev-word
.br
a-b = cursor-prev-word
.br
a-f = cursor-next-word
.br
a-d = delete-next-word
.br
c-v = clipboard-paste
.br
c-p = history-prev
.br
c-n = history-next
.br
c-r = history-search
.br
c-up = history-prev
.br
c-down = history-next

.br
[GntTree::binding]
.br
up = move-up
.br
down = move-down
.br
c-n = move-down
.br
c-p = move-up
.br
pageup = page-up
.br
pagedown = page-down
.br
backspace = move-parent
.br
home = move-first
.br
end = move-last
.br
# Following is the default binding for the context-menu
.br
menu = context-menu
.br
# The following will let you open the context-menu in the buddylist with c-b
.br
# c-b = context-menu

.br
[GntWidget::binding]
.br
f11 = context-menu
.br
c-x = context-menu

[GntWindow::binding]
.br
c-o = show-menu
.br
f10 = show-menu

The \fBc-\fR corresponds to the \fBControl\fR key. You can also use \fBctrl-\fR
or \fBctr-\fR or \fBctl-\fR to indicate a combination. For alt-keys, you can use
one of \fBa-\fR, \fBalt-\fR, \fBm-\fR or \fBmeta-\fR. You can also use
\fBhome\fR, \fBend\fR, \fBleft\fR, \fBright\fR etc. keys.

To unbind a key which has a default binding, you simply bind it to the empty string.  For example, to unbind \fBAlt + q\fR from the Quit function, you would use:

[GntWM::binding]
.br
a-q =

.SH Menus
You can also specify key-bindings to trigger specific menuitems in windows. For example, the following entry in \fI~/.gntrc\fR will bind \fBCtrl + t\fR to the 'Send IM...' item in the buddylist:

[buddylist::menu]
.br
c-t = send-im

The following is the list of IDs of the current menuitems in the buddylist:

send-im
.br
join-chat
.br
show-empty-groups
.br
show-offline-buddies
.br
sort-status
.br
sort-alpha
.br
sort-log
.br
add-buddy
.br
add-chat
.br
add-group

.SH Mouse Support
There is experimental mouse support. You can focus windows, activate buttons,
select rows in a list, scroll using the wheel-scroll etc. Selecting text in a
text-view copies it to the gnt clipboard. Mouse support is disabled by default,
so you need to enable it in \fI~/.gntrc\fR (see the sample above).

.SH Window Management
The default window management is very limited. But it is possible to write
custom window managers to suit your needs. There is a sample window-manager
included (named \fIs.so\fR) which adds a little 'close-button' for the windows,
removes the borders from the buddylist and pops up new windows in the middle of
the screen, instead of at the upper-left corder. It is provided as a sample
simple manager, but it should be possible to write more complex managers, and
it's very possible that the window-manager API will need to be enhanced. Look at
the sample \fI~/.gntrc\fR file above to see how to select a window-manager.

It is also possible to rebind the window-manager actions in \fI~/.gntrc\fR, eg:

.br
[GntWM::binding]
.br
a-n = window-next
.br
a-p = window-prev
.br
a-c = window-close
.br
a-w = window-list
.br
a-d = dump-screen
.br
a-, = shift-left
.br
a-. = shift-right
.br
a-a = action-list
.br
a-m = start-move
.br
a-r = start-resize
.br
a-q = wm-quit
.br
a-l = refresh-screen
.br
a-s = workspace-list
.br
a-t = window-tag
.br
a-T = place-tagged
.br
a-C = toggle-clipboard
.br
a-/ = help-for-widget
.br
a-c-j = window-scroll-down
.br
a-c-k = window-scroll-up
.br
# The following action is still incomplete, and doesn't have a default binding
.br
# switch-window-n
.br
# Other actions: window-next-urgent, window-prev-urgent

# For the sample custom window manager
.br
[GntS::binding]
.br
a-b = toggle-buddylist

# For the irssi window manager
.br
[Irssi::binding]
.br
a-L = move-right
.br
a-H = move-left
.br
a-J = move-down
.br
a-K = move-up

.SH Conversation Commands
There are a few helpful commands in addition to the regular commands. You can
use these from any conversation to access other windows.

.TP
.B accounts
for the accounts window.

.TP
.B buddylist
for the buddylist.

.TP
.B debugwin
for the debug window.

.TP
.B plugins
for the plugins window.

.TP
.B prefs
for the preferences window.

.TP
.B status
for the status window.

.SH FAQ
FAQ for \fBfinch\fR is located at
.br
\fIhttps://developer.pidgin.im/wiki/Using%20Finch\fR

.SH BUGS
Known bugs are listed at
.br
\fIhttps://developer.pidgin.im/query?status=new&status=assigned&status=reopened&component=finch+%28gnt%2Fncurses%29&order=priority\fR

Before sending a bug report, please verify that you have the latest
version of \fBfinch\fR and libpurple.  Many bugs (major and minor) are
fixed at each release, and if yours is out of date, the problem may already have
been solved.

.SH PATCHES
If you fix a bug in \fBfinch\fR (or otherwise enhance it), please submit a
patch (using \fBmtn diff > my.diff\fR against the latest version from the
Mercurial repository) at \fIhttps://developer.pidgin.im/newticket\fR

You are also encouraged to drop by at \fB#pidgin\fR on \fIirc.freenode.net\fR
to discuss development.

.SH SEE ALSO
\fIhttps://pidgin.im/\fR
.br
\fIhttps://developer.pidgin.im/\fR
.br
\fBpurple-remote\fR(1)
.br
\fBpidgin\fR(1)

.SH LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, but
\fBWITHOUT ANY WARRANTY\fR; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111-1301  USA

.SH AUTHORS
Sadrul Habib Chowdhury <\fIsadrul@users.sourceforge.net\fR>
.br

This manpage was written by Sadrul Habib Chowdhury
<\fIsadrul@users.sourceforge.net\fR> and Dennis Ristuccia
<\fIdennis@dennisr.net\fR>.



NEWS
Pidgin and Finch: The Pimpin' Penguin IM Clients That're Good for the Soul

Our development blog is available at: https://planet.pidgin.im

2.8.0 (06/07/2011):
	Paul: I fixed a few things in this release, and committed some
	patches, but am being vague because I can't remember any specifics.
	Hopefully you enjoy this, and happy May!

	John: Wow, this release has taken forever.  But there are a crapload
	of patches that have been committed this time around, a lot of them by
	me.  The infamous MSN bugs are still outstanding, but otherwise this
	release has brought some excellent progress.  Oh, and we dropped the
	hasn't-quite-worked-in-ages QQ plugin, too.  Enjoy!

	Jorge: I managed to get one patch applied! Wii for end of semester!

2.7.11 (03/10/2011):
	John: Yet another release.  This time around we finally fixed that
	annoying MSN buddy adding problem and a security issue in Yahoo.  You
	know the drill--upgrade already!

2.7.10 (02/06/2011):
	John: It's release time again.  This release contains a bunch of stuff
	committed from Trac.  This is another "thank a patch writer" release.
	Unfortunately, no one has fixed our wonderful MSN issues yet.  There is
	a tiny security fix in this release, as well.

2.7.9 (12/26/2010):
	John: Just a quick release for a security fix here.  Elliott has not
	yet had a chance to work on the MSN breakage that's been present in
	the last couple releases, but we hope he can do it before 2.7.10!

2.7.8 (12/19/2010):
	Elliott: OK, so I know a few things broke with the last release, and
	it's too bad we had to rush it for that silly certificate thing that
	the MSN people can't configure properly.  I've certainly done a lot of
	small fixes this time, but it's too bad we haven't been able to get the
	transfers with the official client fixed yet.  I promise it'll be in
	the next release (barring any quick security issues).

	John: So, it's been about a month since we last released.  Again, we've
	assembled a bugfix release for your enjoyment.  While a few commonly
	reported bugs remain, particularly in MSN, we're working on it for the
	next release.  In the meantime, Merry Christmas and enjoy!

2.7.7 (11/23/2010):
	John: Well, this time around, we should finally have the certificate
	issue really and fully fixed for all of you MSN users.  Also, we have
	a few AIM-related fixes in this release, most notably the fix for the
	new "SSL Handshake Failure" message some of you got after upgrading.
	That one was an oversight on our part.  Enjoy the fixes!

2.7.6 (11/21/2010):
	Jorge: In this release I have merged two branches where I have spent
	most of my time in the last months, the MSNP16 and SLP-rewrite.  I 
	hope you all will enjoy the hability to be connected on multiple
	instances at the same time. I also hope that the SLP rewrite
	fix a lot of old bugs that we have in the tracker. I am really 
	happy with this rewrite because there was untoched code from almost
	5 years ago. I hope you like this release!

	John: In this release, we give you some new features and a bunch of
	bug fixes.  This includes shipping intermediate certificates to fix
	certificate validation for MSN's servers.  Upgrade and enjoy!

2.7.5 (10/31/2010):
	John: A bugfix release for all of you!  This time we fixed a bunch of
	bugs ranging from annoying regressions to long-standing bugs we didn't
	realize until now were bugs.  Enjoy!

2.7.4 (10/20/2010):
	John: This release came at this particular time due to some security
	issues Daniel discovered for us when investigating a bug.  There are
	a ton of other changes, including some partial Yahoo file transfer
	fixes and a bunch of other little things.  Enjoy!

2.7.3 (08/10/2010):
	Mark: Lots of little incremental[1] bug fixes and enhancements in this
	release.

	[1] No whales were harmed[2] during the creation of this release.
	[2] Probably.

	John: Finally got some fixes out there for you Yahoo users behind some
	particularly annoying firewalls and proxies, among other fixes.  Enjoy!

2.7.2 (07/21/2010):
	Mark: We discovered a security issue in Pidgin 2.7.0 and 2.7.1 and
	decided to release a patched version quickly.  This release contains
	the fix for that crash, and a few other minor fixes.

2.7.1 (05/29/2010):
	Elliott: Hey, I'm first!  How did that happen?!  Maybe because of the
	interesting changes in this release.  Sure there were quite a few bug
	fixes, but I know what you've all been waiting for is the direct
	connections in MSN.  Trust me, it's really really fast!!

	John: Whoa, short turnaround for us.  This is just 17 days after our
	previous release!  This fixes a number of bugs that you've all been
	reporting a ton of duplicate tickets about and even gives you the new
	direct connection file transfer support for MSN.  Enjoy!

	Marcus: Quite a bit quicker to get this release out, compared with the
	previous one :).  Fixes a number of bugs, and I'm sure the MSN direct
	connections will please many users.  Enjoy!

2.7.0 (05/12/2010):
	John: We FINALLY got the ICQ X-Status stuff merged in!  And a few other
	patches that have been sitting on Trac forever.  Couple that with some
	new features and we have an excellent release for all of you!

	Marcus: Finally time for a new release, seems like it took
	quite a while this time.  But then again, there are some new features
	in there, like file transfer preview (thumbnails) support (only on MSN
	so far), plus a lot of bug fixes.  Enjoy!

	Paul: Yay, a new release!  I don't think I added very much useful to
	this release, other than fixing a paste bug from Chrome (no weird
	characters appended to the end of your URI).  Enjoy!

	Jorge: This is my first NEWS!  I'm not sure I added something nice
	to this release.  I know it took a lot of time to bring this one
	out, however I'm really excited by this release.  I hope everyone
	likes the new features this release brings.

	Elliott: This release took so long, I had to go check the ChangeLog

finch/finch.c
/**
 * finch
 *
 * Finch is the legal property of its developers, whose names are too numerous
 * to list here.  Please refer to the COPYRIGHT file distributed with this
 * source distribution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111-1301  USA
 */
#include <internal.h>
#include "finch.h"

#include "account.h"
#include "conversation.h"
#include "core.h"
#include "debug.h"
#include "eventloop.h"
#include "ft.h"
#include "log.h"
#include "notify.h"
#include "prefs.h"
#include "prpl.h"
#include "pounce.h"
#include "savedstatuses.h"
#include "sound.h"
#include "status.h"
#include "util.h"
#include "whiteboard.h"

#include "gntdebug.h"
#include "gntprefs.h"
#include "gntui.h"
#include "gntidle.h"

#define _GNU_SOURCE
#include <getopt.h>

#include "config.h"
#include "package_revision.h"

static void
debug_init(void)
{
	finch_debug_init();
	purple_debug_set_ui_ops(finch_debug_get_ui_ops());
}

static GHashTable *ui_info = NULL;
static GHashTable *finch_ui_get_info(void)
{
	if (ui_info == NULL) {
		ui_info = g_hash_table_new(g_str_hash, g_str_equal);

		g_hash_table_insert(ui_info, "name", (char*)_("Finch"));
		g_hash_table_insert(ui_info, "version", VERSION);
		g_hash_table_insert(ui_info, "website", "http://pidgin.im");
		g_hash_table_insert(ui_info, "dev_website", "http://developer.pidgin.im");
		g_hash_table_insert(ui_info, "client_type", "console");

		/*
		 * This is the client key for "Finch." Please don't use this
		 * key for other applications.  You can not specify a client
		 * key, in which case the default "libpurple" key will be used
		 */
		g_hash_table_insert(ui_info, "prpl-aim-clientkey", "ma18nmEklXMR7Cj_");

		/*
		 * This is the client key for "Pidgin."  It is owned by the AIM
		 * account "markdoliner."  Please don't use this key for other
		 * applications.  You can either not specify a client key, in
		 * which case the default "libpurple" key will be used, or you
		 * can try to register your own at the AIM or ICQ web sites
		 * (although this functionality was removed at some point, it's
		 * possible it has been re-added).  AOL's old key management
		 * page is http://developer.aim.com/manageKeys.jsp
		 *
		 * We used to have a Finch-specific devId/clientkey
		 * (ma19sqWV9ymU6UYc), but it stopped working, so we switched
		 * to this one.
		 */
		g_hash_table_insert(ui_info, "prpl-icq-clientkey", "ma1cSASNCKFtrdv9");

		/*
		 * This is the distid for Finch, given to us by AOL.  Please
		 * don't use this for other applications.  You can just not
		 * specify a distid and libpurple will use a default.
		 */
		g_hash_table_insert(ui_info, "prpl-aim-distid", GINT_TO_POINTER(1718));
		g_hash_table_insert(ui_info, "prpl-icq-distid", GINT_TO_POINTER(1552));
	}

	return ui_info;
}

static void
finch_quit(void)
{
	gnt_ui_uninit();
	if (ui_info)
		g_hash_table_destroy(ui_info);
}

static PurpleCoreUiOps core_ops =
{
	finch_prefs_init,
	debug_init,
	gnt_ui_init,
	finch_quit,
	finch_ui_get_info,

	/* padding */
	NULL,
	NULL,
	NULL
};

static PurpleCoreUiOps *
gnt_core_get_ui_ops(void)
{
	return &core_ops;
}

/* Anything IO-related is directly copied from gtkpurple's source tree */

#define FINCH_READ_COND  (G_IO_IN | G_IO_HUP | G_IO_ERR)
#define FINCH_WRITE_COND (G_IO_OUT | G_IO_HUP | G_IO_ERR | G_IO_NVAL)

typedef struct _PurpleGntIOClosure {
	PurpleInputFunction function;
	guint result;
	gpointer data;

} PurpleGntIOClosure;

static void purple_gnt_io_destroy(gpointer data)
{
	g_free(data);
}

static gboolean purple_gnt_io_invoke(GIOChannel *source, GIOCondition condition, gpointer data)
{
	PurpleGntIOClosure *closure = data;
	PurpleInputCondition purple_cond = 0;

	if (condition & FINCH_READ_COND)
		purple_cond |= PURPLE_INPUT_READ;
	if (condition & FINCH_WRITE_COND)
		purple_cond |= PURPLE_INPUT_WRITE;

#if 0
	purple_debug(PURPLE_DEBUG_MISC, "gtk_eventloop",
			   "CLOSURE: callback for %d, fd is %d\n",
			   closure->result, g_io_channel_unix_get_fd(source));
#endif

#ifdef _WIN32
	if(! purple_cond) {
#if DEBUG
		purple_debug_misc("gnt_eventloop",
			   "CLOSURE received GIOCondition of 0x%x, which does not"
			   " match 0x%x (READ) or 0x%x (WRITE)\n",
			   condition, FINCH_READ_COND, FINCH_WRITE_COND);
#endif /* DEBUG */

		return TRUE;
	}
#endif /* _WIN32 */

	closure->function(closure->data, g_io_channel_unix_get_fd(source),
			  purple_cond);

	return TRUE;
}

static guint gnt_input_add(gint fd, PurpleInputCondition condition, PurpleInputFunction function,
							   gpointer data)
{
	PurpleGntIOClosure *closure = g_new0(PurpleGntIOClosure, 1);
	GIOChannel *channel;
	GIOCondition cond = 0;

	closure->function = function;
	closure->data = data;

	if (condition & PURPLE_INPUT_READ)
		cond |= FINCH_READ_COND;
	if (condition & PURPLE_INPUT_WRITE)
		cond |= FINCH_WRITE_COND;

	channel = g_io_channel_unix_new(fd);
	closure->result = g_io_add_watch_full(channel, G_PRIORITY_DEFAULT, cond,
					      purple_gnt_io_invoke, closure, purple_gnt_io_destroy);

	g_io_channel_unref(channel);
	return closure->result;
}

static PurpleEventLoopUiOps eventloop_ops =
{
	g_timeout_add,
	g_source_remove,
	gnt_input_add,
	g_source_remove,
	NULL, /* input_get_error */
#if GLIB_CHECK_VERSION(2,14,0)
	g_timeout_add_seconds,
#else
	NULL,
#endif

	/* padding */
	NULL,
	NULL,
	NULL
};

static PurpleEventLoopUiOps *
gnt_eventloop_get_ui_ops(void)
{
	return &eventloop_ops;
}

/* This is mostly copied from gtkpurple's source tree */
static void
show_usage(const char *name, gboolean terse)
{
	char *text;

	if (terse) {
		text = g_strdup_printf(_("%s. Try `%s -h' for more information.\n"), DISPLAY_VERSION, name);
	} else {
		text = g_strdup_printf(_("%s\n"
		       "Usage: %s [OPTION]...\n\n"
		       "  -c, --config=DIR    use DIR for config files\n"
		       "  -d, --debug         print debugging messages to stderr\n"
		       "  -h, --help          display this help and exit\n"
		       "  -n, --nologin       don't automatically login\n"
		       "  -v, --version       display the current version and exit\n"), DISPLAY_VERSION, name);
	}

	purple_print_utf8_to_console(stdout, text);
	g_free(text);
}

static int
init_libpurple(int argc, char **argv)
{
	char *path;
	int opt;
	gboolean opt_help = FALSE;
	gboolean opt_nologin = FALSE;
	gboolean opt_version = FALSE;
	char *opt_config_dir_arg = NULL;
	gboolean debug_enabled = FALSE;

	struct option long_options[] = {
		{"config",   required_argument, NULL, 'c'},
		{"debug",    no_argument,       NULL, 'd'},
		{"help",     no_argument,       NULL, 'h'},
		{"nologin",  no_argument,       NULL, 'n'},
		{"version",  no_argument,       NULL, 'v'},
		{0, 0, 0, 0}
	};

#ifdef ENABLE_NLS
	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "UTF-8");
	textdomain(PACKAGE);
#endif

#ifdef HAVE_SETLOCALE
	setlocale(LC_ALL, "");
#endif

	/* scan command-line options */
	opterr = 1;
	while ((opt = getopt_long(argc, argv,
#ifndef _WIN32
				  "c:dhn::v",
#else
				  "c:dhn::v",
#endif
				  long_options, NULL)) != -1) {
		switch (opt) {
		case 'c':	/* config dir */
			g_free(opt_config_dir_arg);
			opt_config_dir_arg = g_strdup(optarg);
			break;
		case 'd':	/* debug */
			debug_enabled = TRUE;
			break;
		case 'h':	/* help */
			opt_help = TRUE;
			break;
		case 'n':	/* no autologin */
			opt_nologin = TRUE;
			break;
		case 'v':	/* version */
			opt_version = TRUE;
			break;
		case '?':	/* show terse help */
		default:
			show_usage(argv[0], TRUE);
			return 0;
			break;
		}
	}

	/* show help message */
	if (opt_help) {
		show_usage(argv[0], FALSE);
		return 0;
	}
	/* show version message */
	if (opt_version) {
		/* Translators may want to transliterate the name.
		 It is not to be translated. */
		printf("%s %s (%s)\n", _("Finch"), DISPLAY_VERSION, REVISION);
		return 0;
	}

	/* set a user-specified config directory */
	if (opt_config_dir_arg != NULL) {
		purple_util_set_user_dir(opt_config_dir_arg);
		g_free(opt_config_dir_arg);
	}

	/*
	 * We're done piddling around with command line arguments.
	 * Fire up this baby.
	 */

	/* We don't want debug-messages to show up and corrupt the display */
	purple_debug_set_enabled(debug_enabled);

	/* If we're using a custom configuration directory, we
	 * do NOT want to migrate, or weird things will happen. */
	if (opt_config_dir_arg == NULL)
	{
		if (!purple_core_migrate())
		{
			char *old = g_strconcat(purple_home_dir(),
			                        G_DIR_SEPARATOR_S ".gaim", NULL);
			char *text = g_strdup_printf(_(
				"%s encountered errors migrating your settings "
				"from %s to %s. Please investigate and complete the "
				"migration by hand. Please report this error at http://developer.pidgin.im"), _("Finch"),
				old, purple_user_dir());

			g_free(old);

			purple_print_utf8_to_console(stderr, text);
			g_free(text);

			return 0;
		}
	}

	purple_core_set_ui_ops(gnt_core_get_ui_ops());
	purple_eventloop_set_ui_ops(gnt_eventloop_get_ui_ops());
	purple_idle_set_ui_ops(finch_idle_get_ui_ops());

	path = g_build_filename(purple_user_dir(), "plugins", NULL);
	if (g_mkdir(path, S_IRUSR | S_IWUSR | S_IXUSR) != 0 && errno != EEXIST)
		fprintf(stderr, "Couldn't create plugins dir\n");
	purple_plugins_add_search_path(path);
	g_free(path);

	purple_plugins_add_search_path(LIBDIR);

	if (!purple_core_init(FINCH_UI))
	{
		fprintf(stderr,
				"Initialization of the Purple core failed. Dumping core.\n"
				"Please report this!\n");
		abort();
	}

	/* TODO: Move blist loading into purple_blist_init() */
	purple_set_blist(purple_blist_new());
	purple_blist_load();

	/* TODO: should this be moved into finch_prefs_init() ? */
	finch_prefs_update_old();

	/* load plugins we had when we quit */
	purple_plugins_load_saved("/finch/plugins/loaded");

	/* TODO: Move pounces loading into purple_pounces_init() */
	purple_pounces_load();

	if (opt_nologin)
	{
		/* Set all accounts to "offline" */
		PurpleSavedStatus *saved_status;

		/* If we've used this type+message before, lookup the transient status */
		saved_status = purple_savedstatus_find_transient_by_type_and_message(
							PURPLE_STATUS_OFFLINE, NULL);

		/* If this type+message is unique then create a new transient saved status */
		if (saved_status == NULL)
			saved_status = purple_savedstatus_new(NULL, PURPLE_STATUS_OFFLINE);

		/* Set the status for each account */
		purple_savedstatus_activate(saved_status);
	}
	else
	{
		/* Everything is good to go--sign on already */
		if (!purple_prefs_get_bool("/purple/savedstatus/startup_current_status"))
			purple_savedstatus_activate(purple_savedstatus_get_startup());
		purple_accounts_restore_current_statuses();
	}

	return 1;
}

static gboolean gnt_start(int *argc, char ***argv)
{
	/* Initialize the libpurple stuff */
	if (!init_libpurple(*argc, *argv))
		return FALSE;

	purple_blist_show();
	return TRUE;
}

int main(int argc, char *argv[])
{
	signal(SIGPIPE, SIG_IGN);

#if !GLIB_CHECK_VERSION(2, 32, 0)
	/* GLib threading system is automaticaly initialized since 2.32.
	 * For earlier versions, it have to be initialized before calling any
	 * Glib or GTK+ functions.
	 */
	g_thread_init(NULL);
#endif

	g_set_prgname("Finch");
	g_set_application_name(_("Finch"));

	if (gnt_start(&argc, &argv)) {
		gnt_main();

#ifdef STANDALONE
		purple_core_quit();
#endif
	}

	return 0;
}

